console/src/main/java/com/alibaba/nacos/console/controller/OidcAuthController.java

```diff
@@ -0,0 +1,290 @@
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.net.SocketTimeoutException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * OIDC auth controller.
+ *
+ * @author Kicey
+ */
+@RestController
+@RequestMapping("/v1/auth/oidc")
+public class OidcAuthController {
+    
+    @Autowired
+    private OidcService oidcService;
+    
+    @Autowired
+    private NacosUserDetailsServiceImpl userDetailsService;
+    
+    /**
+     * List all OIDC IDP configurations.
+     *
+     * @return OIDC IDP configurations
+     */
+    @GetMapping("/list")
+    public List<Map<String, String>> list() {
+        List<Map<String, String>> oidpList = new ArrayList<>();
+        List<String> oidpListFromConfig = OidcUtil.getOidpList();
+        oidpListFromConfig.forEach(oidp -> {
+            Map<String, String> curOidp = new HashMap<>(4);
+            curOidp.put("key", oidp);
+            curOidp.put("name", OidcUtil.getName(oidp));
+            oidpList.add(curOidp);
+        });
+        return oidpList;
+    }
+    
+    /**
+     * Initiate OIDC authentication flow.
+     * A state object will be initialized here, which will be base64-encoded, passed to the IDP and eventually relayed to the callback endpoint.
+     *
+     * @param id       id of the IDP
+     * @param origin   origin of the request from browser
+     * @param response response object
+     */
+    @GetMapping("/init")
+    public Object init(@RequestParam("id") String id, @RequestParam("origin") String origin, HttpServletResponse response) {
+        
+        String authUrl = OidcUtil.getAuthUrl(id);
+        if (authUrl == null) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND)
+                    .body(String.format("OIDC IDP %s not found in config", id));
+        }
+        
+        String clientId = OidcUtil.getClientId(id);
+        
+        List<String> scopes = OidcUtil.getScopes(id);
+
+        String host = OidcUtil.getExposedHost();
+        if (host == null) {
+            host = origin;
+        }
+        String redirectUrl = host + OidcUtil.CALLBACK_PATH;
+        
+        String state;
```

Comment:
The state should be random value for every time oauth2 authentication to avoid csrf. We can not set the unchanged values.

Reply:
Indeed the state should be random according to the oidc standard, however to achieve this we must save the state to another place at the init endpoint, in order to verify it at the callback endpoint. I think the only option was saving it to the cookie, but nacos console has no any code using the cookie before, so I'm afraid of causing some compatiblity problems. Is it acceptible to use cookie here?

Reply:
Yes. Currently, it is the most common way to store state value using cookie. Introducing other persistent store server is complex and heavy. I recommend to use cookie.
Today, for the sake of security, we should follow the standard protocol even if we need to modify more codes.

Reply:
store the random number(Long type) in the cookie to check the state